### PR TITLE
add map & filter functions

### PR DESCRIPTION
**Pull request by Yexiaoxing**
reduce function will be added later

### CODE REVIEW

File: src/include/algorithm/Filter.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+//
+// Created by Xiaoxing Ye on 08/05/2017.
+//
+
+#ifndef CUHKSZ_ALGORITHM_FILTER
+#define CUHKSZ_ALGORITHM_FILTER
+namespace cuhksz {
+
+template<typename Container, typename Func>
+std::vector<typename Container::value_type> filter(Container &cont, Func func) {
```
Comment (alesiong) [inline src/include/algorithm/Filter.h]: Seems that the containers in our lib do not have `value_type`

File: src/include/algorithm/Map.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
+//
+// Created by Xiaoxing Ye on 08/05/2017.
+//
+
+#ifndef CUHKSZ_ALGORITHM_MAP
+#define CUHKSZ_ALGORITHM_MAP
+namespace cuhksz {
+template<typename Container, typename Func>
+std::vector<typename std::result_of<Func(typename Container::value_type &)>::type> map(Container &cont, Func func) {
```
Comment (alesiong) [inline src/include/algorithm/Map.h]: consider name conflict if both using `std` and `cuhksz`

**Comment by alesiong** [comment]
Please include needed header in `Map.h` and `Filter.h`

**Comment by alesiong** [comment]
In fact `std` has `transform` and `accumulate`

File: src/include/algorithm/Filter.h (hunk 1 above)
Comment (Yexiaoxing) [inline src/include/algorithm/Filter.h]: do we need to add it?

**Comment by alesiong** [comment]
Good, should we ask @Larry-u to add `value_type` for our container class?

**Comment by Yexiaoxing** [comment]
I think we should

**Pull request merged**